File: arthur/backend/media/ArthurNumber.java
package arthur.backend.media;

/**
 * Java implementation of arthur number!
 */
public class ArthurNumber extends ArthurMedia implements java.io.Serializable {

  public static final String NUMBER = "number";

  public Double val;

  public ArthurNumber() {
    this(0.0);
  }

  public ArthurNumber(double val) {
    this.type = NUMBER;
    this.val = val;
  }

  public ArthurNumber add(ArthurMedia two) {
    if (two.type.equals(NUMBER)) {
      return JavaNumberMath.add(this, (ArthurNumber) two);
    } else {
      ArthurNumber num = (ArthurNumber) two.castTo("num");
      return JavaNumberMath.divide(this, num);
    }
  }

  public ArthurNumber minus(ArthurMedia two) {
    if (two.type.equals(NUMBER)) {
      return JavaNumberMath.minus(this, (ArthurNumber) two);
    } else {
      ArthurNumber num = (ArthurNumber) two.castTo("num");
      return JavaNumberMath.divide(this, num);
    }
  }

  public ArthurNumber multiply(ArthurMedia two) {
    if (two.type.equals(NUMBER)) {
      return JavaNumberMath.multiply(this, (ArthurNumber) two);
    } else {
      ArthurNumber num = (ArthurNumber) two.castTo("num");
      return JavaNumberMath.divide(this, num);
    }
  }

  public ArthurNumber divide(ArthurMedia two) {
    if (two.type.equals(NUMBER)) {
      return JavaNumberMath.divide(this, (ArthurNumber) two);
    } else {
      ArthurNumber num = (ArthurNumber) two.castTo("num");
      return JavaNumberMath.divide(this, num);
    }
  }

  public boolean lessThan(ArthurMedia two) {
    if (two.type.equals(NUMBER)) {
      ArthurNumber t = (ArthurNumber) two;
      return (this.val < t.val);
    } else {
      ArthurNumber num = (ArthurNumber) two.castTo("num");
      return lessThan(num);
    }
  }

  public boolean greaterThan(ArthurMedia two) {
    if (two.type.equals(NUMBER)) {
      ArthurNumber t = (ArthurNumber) two;
      return (this.val > t.val);
    } else {
      ArthurNumber num = (ArthurNumber) two.castTo("num");
      return greaterThan(num);
    }
  }

  public boolean lessThanEquals(ArthurMedia two) {
    if (two.type.equals(NUMBER)) {
      ArthurNumber t = (ArthurNumber) two;
      if (this.val < t.val || this.arthurEquals(two))
        return true;
      else
        return false;
    } else {
      ArthurNumber num = (ArthurNumber) two.castTo("num");
      return lessThanEquals(num);
    }
  }

  public boolean greaterThanEquals(ArthurMedia two) {
    if (two.type.equals(NUMBER)) {
      ArthurNumber t = (ArthurNumber) two;
      if (this.val > t.val || this.arthurEquals(two))
        return true;
      else
        return false;
    } else {
      ArthurNumber num = (ArthurNumber) two.castTo("num");
      return greaterThanEquals(num);
    }
  }

  public boolean arthurEquals(ArthurMedia two) {
    if (two.type.equals(NUMBER)) {
      ArthurNumber t = (ArthurNumber) two;
      double diff = Math.abs(this.val - t.val);
      return (diff <= 0.0001);
    } else {
      ArthurNumber num = (ArthurNumber) two.castTo("num");
      return arthurEquals(num);
    }
  }

  public ArthurMedia castTo(ArthurString mediaType) {
    return castTo(mediaType.str);
  }

  public ArthurMedia castTo(String mediaType) {
    if (mediaType.equals("Image")) {
      return this.toImage();
    } else if (mediaType.equals("Sound")) {
      return this.toSound();
    } else if (mediaType.equals("string")) {
      return this.toArtString();
    } else if (mediaType.equals("Video")) {
      return this.toVideo();
    } else if (mediaType.equals("color")) {
      return this.toColor();
    }

    return this;
  }

  public ArthurColor toColor() {
    double r = Math.max(this.val, 255);
    double g = Math.max(this.val, 255);
    double b = Math.max(this.val, 255);
    double a = Math.max(this.val / 255.0, 1.0);
    return new ArthurColor(r, g, b, a);
  }

  public ArthurString toArtString() {
    String str = "" + this.val;
    return new ArthurString(str);
  }

  public ArthurVideo toVideo() {
    ArthurVideo zero = new ArthurVideo(ArthurVideo.ZERO);
    return zero.add(this);
  }

  public ArthurImage toImage() {
    ArthurImage zero = new ArthurImage(ArthurImage.ZERO);
    return zero.add(this);
  }

  public ArthurSound toSound() {
    ArthurSound zero = new ArthurSound(ArthurSound.ZERO);
    return zero.add(this);
  }

  public int intval() {
    return (int) Math.round(this.val);
  }

  public String toString() {
    return "" + this.val;
  }

  public String jsLiteral() {
    return "new ArthurNumber(" + this.val + ")";
  }

}

File: samples/sample16trans/js/arthur/arthur-number.js

var types = require('./types');
var ArthurMedia = require('./arthur-media');
var ArthurColor, ArthurString;

module.exports = ArthurNumber;

ArthurNumber.prototype.__proto__ = ArthurMedia.prototype;

function ArthurNumber(num) {
  if (typeof num === 'object') {
    return new ArthurNumber(num.val);
  }

  this.type = types.NUMBER;
  this.val = parseFloat(num) || 0;
}

ArthurNumber.prototype.int = function() {
  return Math.round(this.val);
}

ArthurNumber.prototype.add = function(num) {
  if (num.type && num.type == types.NUMBER) {
    return new ArthurNumber(this.val + num.val);
  }

  if (num.type && (num.type == types.STRING || num.type == types.COLOR)) {
    var s = num.castTo('num');
    return this.add(s);
  }

  return this;
}

ArthurNumber.prototype.minus = function(num) {
  if (num.type && num.type == types.NUMBER) {
    return new ArthurNumber(this.val - num.val);
  }

  if (num.type && (num.type == types.STRING || num.type == types.COLOR)) {
    var s = num.castTo('num');
    return this.minus(s);
  }

  return this;
}

ArthurNumber.prototype.multiply = function(num) {
  if (num.type && num.type == types.NUMBER) {
    return new ArthurNumber(this.val * num.val);
  }

  if (num.type && (num.type == types.STRING || num.type == types.COLOR)) {
    var s = num.castTo('num');
    return this.multiply(s);
  }

  return this;
}

ArthurNumber.prototype.divide = function(num) {
  if (num.type && num.type == types.NUMBER) {
    return new ArthurNumber(this.val / num.val);
  }

  if (num.type && (num.type == types.STRING || num.type == types.COLOR)) {
    var s = num.castTo('num');
    return this.divide(s);
  }

  return this;
}

ArthurNumber.prototype.castTo = function(t) {
  if (typeof ArthurColor != 'function') {
    ArthurColor = require('./arthur-color');
  }
  if (typeof ArthurString != 'function') {
    ArthurString = require('./arthur-string');
  }

  if (t == 'color') {
    var r = Math.max(this.val, 255);
    var g = Math.max(this.val, 255);
    var b = Math.max(this.val, 255);
    var a = Math.max(this.val / 255.0, 1.0);
    return new ArthurColor(r, g, b, a);
  }

  if (t == 'string') {
    return new ArthurString('' + this.val);
  }

  return this;
}

ArthurNumber.prototype.lessThan = function(num) {
  return (this.val < num.val);
}

ArthurNumber.prototype.lessThanEquals = function(num) {
  return (this.val <= num.val);
}

ArthurNumber.prototype.greaterThan = function(num) {
  return (this.val > num.val);
}

ArthurNumber.prototype.greaterThanEquals = function(num) {
  return (this.val >= num.val);
}

ArthurNumber.prototype.arthurEquals = function(num) {
  return (this.int() == num.int());
}
